clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -825,6 +825,97 @@ protected void produceToLocalKafka(
     }
   }
 
+  @Override
+  protected Map<String, Long> calculateLeaderUpstreamOffsetWithTopicSwitch(
+      PartitionConsumptionState partitionConsumptionState,
+      TopicSwitch topicSwitch,
+      PubSubTopic newSourceTopic,
+      List<CharSequence> unreachableBrokerList) {
+    Map<String, Long> upstreamOffsetsByKafkaURLs = new HashMap<>(topicSwitch.sourceKafkaServers.size());
+    final PubSubTopicPartition sourceTopicPartition = partitionConsumptionState.getSourceTopicPartition(newSourceTopic);
+    long rewindStartTimestamp;
+    // calculate the rewind start time here if controller asked to do so by using this sentinel value.
+    if (topicSwitch.rewindStartTimestamp == REWIND_TIME_DECIDED_BY_SERVER) {
+      rewindStartTimestamp = calculateRewindStartTime(partitionConsumptionState);
+      LOGGER.info(
+          "{} leader calculated rewindStartTimestamp {} for {}",
+          ingestionTaskName,
+          rewindStartTimestamp,
+          sourceTopicPartition);
+    } else {
+      rewindStartTimestamp = topicSwitch.rewindStartTimestamp;
+    }
+
+    topicSwitch.sourceKafkaServers.forEach(sourceKafkaURL -> {
+      Long upstreamStartOffset =
+          partitionConsumptionState.getLatestProcessedUpstreamRTOffsetWithNoDefault(sourceKafkaURL.toString());
+      if (upstreamStartOffset == null || upstreamStartOffset < 0) {
+        if (rewindStartTimestamp > 0) {
+          PubSubTopicPartition newSourceTopicPartition =
+              new PubSubTopicPartitionImpl(newSourceTopic, sourceTopicPartition.getPartitionNumber());
+          try {
+            upstreamStartOffset =
+                getTopicPartitionOffsetByKafkaURL(sourceKafkaURL, newSourceTopicPartition, rewindStartTimestamp);
+            LOGGER.info(
+                "Get upstreamStartOffset: {} for source URL: {}, topic: {}, rewind timestamp: {}",
```

Comment:
Please add consumption task ID as log prefix; otherwise, since source topic might be RT, we wouldn't know which version this log is for.